ec2net-functions-lib

```diff
@@ -0,0 +1,70 @@
+# -*-Shell-script-*-
+
+# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the MIT License. See the LICENSE accompanying this file
+# for the specific language governing permissions and limitations under
+# the License.
+
+# Library of function shared across various components of
+# ec2-net-utils.  Ideally the functions here would have been added
+# directly to ec2net-functions, and it would have been imported into
+# programs that need it.  However, that file is large, and makes a
+# number of modifications to the global process namespace.  We want to
+# make these functions usable in environments like dhclient hooks,
+# which all run within a single shell process, and modifying the
+# global namespace so extensively is unsafe.  So we introduce this
+# file and move definitions into it that are less likely to conflict
+# with third-party hooks or other such modifications.
+
+# Normally the ec2-net-utils scripts assume ownership of the ifcfg-* configuration files.
+should_sync_interface() {
+    local interface=$1
+    local config_file=$2
+    if [ -z "$interface" ] ||
+           [ -z "$config_file" ]; then
+        return 1
+    fi
+    local unmanaged
+    if [ -s ${config_file} ]; then
+        # make no changes to unmanaged interfaces
+        unmanaged=$(LANG=C grep -l "^[[:space:]]*EC2SYNC=no\([[:space:]#]\|$\)" $config_file)
+        if [ "${config_file}" == "${unmanaged}" ]; then
+            logger --tag ec2net "${interface} should not be unmanaged"
+            return 1
+        fi
+    fi
+    return 0
+}
+
+# VPC assign IPv4 addresses in the 169.254/16 link-local range in
+# IPv6-only subnets.
+subnet_supports_ipv4() {
+    local address=$1
+    case "$address" in
+        "")
+            return 1
+            ;;
+        169.254.*)
+            return 1
+            ;;
+        *)
+            return 0
+            ;;
+    esac
+}
+
+get_interface_rt_table() {
+    local interface=$1
+    # generate a routing table number
+    local RTABLE=${interface#eth}
+    let RTABLE+=10000
+    echo "$RTABLE"
```

Comment:
Can this overflow? What's the largest possible routing table number? Is there a potential for a collision here? What happens if this is a sub-interface?

Reply:
Yes, it wraps around somewhere above 4611686018427387904. I don't believe we'll need to handle this many interfaces on a Linux system.

Note that the logic around handling the route table IDs is inherited from previous versions of net-utils. I'm not inclined to change it here, as it's visible to customers and thus brings added risk. As there hasn't been any issue with this logic (as far as I know, anyway), it would be best not to change it.

As to the handling of sub-interfaces, that's a potentially interesting question. I'm not sure that existing ec2-net-utils releases will handle them gracefully at all right now, and it's certainly not something we actively test.

Reply:
Yeah. This comment was mostly stream of consciousness and the sub-interface bit was what was really bothering me as I was wondering if that became an issue in the presence of a sub-interface.